File: AlgorithmLcr/Try2-Known component types, quantities and parameters/netgraph_id/selector.py
"""Candidate ranking and equivalence clustering (DESIGN.md section 8).

Mirrors the Try1 selector philosophy: rank by RSS (AICc adds a constant
here since every candidate has the same component set), then merge
electrically equivalent candidates -- distinct wirings can share the same
Z(s) exactly (graph symmetries, Whitney 2-isomorphism, value-coincidence
Y-delta transformations), and the correct output of an identification run
is an ordered list of equivalence classes, not a unique "winner".

Equivalence test (Try1 selector): maximum relative difference of Z on a
validation grid (band expanded 10x at each end, 200 log-spaced points)
below max(tol, 3 * sigma_rel_hat).

Secondary order inside a class (Try1 uses fewest elements + plausibility):
fewer internal junctions first, then series-parallel wirings before
bridged ones (Valdes-Tarjan-Puech test), then canonical serialization.
"""

from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np

from .components import ComponentSet
from .graph import Network, is_series_parallel
from .metric import aicc, fit_metrics, rss_of, residual_vector
from .nodal import StructureStamps, network_z


@dataclass
class Candidate:
    """One evaluated candidate network."""

    network: Network
    z_fit: np.ndarray                    # model Z at the measured freqs
    rss: float
    aicc_val: float
    wrmse: float
    max_rel_err: float
    sp: bool = True                      # series-parallel wiring?
    note: str = ""

    @property
    def n_internal(self) -> int:
        return self.network.structure.n_internal


@dataclass
class EquivalenceClass:
    representative: Candidate
    members: list[Candidate] = field(default_factory=list)

    @property
    def rss(self) -> float:
        return self.representative.rss

    @property
    def wrmse(self) -> float:
        return self.representative.wrmse

    @property
    def max_rel_err(self) -> float:
        return self.representative.max_rel_err

    @property
    def aicc(self) -> float:
        return self.representative.aicc_val

    @property
    def n_members(self) -> int:
        return 1 + len(self.members)


def evaluate_candidates(nets: list[Network], compset: ComponentSet,
                        s: np.ndarray, z: np.ndarray, w: np.ndarray,
                        batch_size: int = 4096) -> list[Candidate]:
    """Full-band evaluation of funnel survivors into Candidates."""
    n_obs = 2 * len(z)
    p = compset.n_params
    out: list[Candidate] = []
    by_structure: dict = {}
    for net in nets:
        by_structure.setdefault(net.structure.key, (net.structure, []))[1].append(net.assign)
    for structure, assigns in by_structure.values():
        stamps = StructureStamps.build(structure, compset)
        sp_flag = is_series_parallel(structure.V, structure.mult)
        for i0 in range(0, len(assigns), batch_size):
            chunk = np.array(assigns[i0:i0 + batch_size], dtype=np.intp)
            z_model = stamps.z_full(chunk, s)
            for row, assign in zip(range(chunk.shape[0]), chunk):
                zf = z_model[row]
                if not np.all(np.isfinite(zf)):
                    continue
                res = residual_vector(z, zf, w)
                rss = rss_of(res)
                wrmse, mre = fit_metrics(z, zf)
                out.append(Candidate(
                    network=Network(structure=structure, assign=tuple(int(x) for x in assign)),
                    z_fit=zf, rss=rss, aicc_val=aicc(rss, n_obs, p),
                    wrmse=wrmse, max_rel_err=mre, sp=sp_flag))
    out.sort(key=lambda c: c.rss)
    return out


def make_validation_grid(f: np.ndarray, n: int = 200,
                         expand: float = 10.0) -> np.ndarray:
    """Band expanded `expand`x at each end, `n` log-spaced points (Try1)."""
    fmin, fmax = float(np.min(f)), float(np.max(f))
    return np.logspace(np.log10(fmin / expand), np.log10(fmax * expand), n)


def are_equivalent(net_a: Network, net_b: Network, compset: ComponentSet,
                   grid: np.ndarray, tol: float) -> bool:
    """Try1 numerical equivalence: max_k |Z1 - Z2| / |Z1| < tol on the grid."""
    za = network_z(net_a, compset, grid)
    zb = network_z(net_b, compset, grid)
    return _rel_diff_below(za, zb, tol)


def _rel_diff_below(za: np.ndarray, zb: np.ndarray, tol: float) -> bool:
    denom = np.abs(za)
    tiny = denom < 1e-300
    if np.any(tiny):
        denom = np.where(tiny, np.abs(zb), denom)
    rel = np.abs(za - zb) / denom
    rel = rel[np.isfinite(rel)]
    if len(rel) == 0:
        return False
    return float(np.max(rel)) < tol


def rank_and_cluster(candidates: list[Candidate], compset: ComponentSet,
                     f: np.ndarray, *, cluster_top: int = 50,
                     equiv_tol: float = 1e-3) -> list[EquivalenceClass]:
    """Sort by RSS and cluster the top `cluster_top` candidates."""
    if not candidates:
        return []
    sigma_hat = min(c.wrmse for c in candidates)
    tol = max(equiv_tol, 3.0 * sigma_hat)
    grid = make_validation_grid(f)
    top = candidates[:cluster_top]
    z_grids = [network_z(c.network, compset, grid) for c in top]
    classes: list[tuple[EquivalenceClass, np.ndarray]] = []
    for cand, zg in zip(top, z_grids):
        for cl, z_rep in classes:
            if _rel_diff_below(z_rep, zg, tol):
                cl.members.append(cand)
                break
        else:
            classes.append((EquivalenceClass(representative=cand), zg))
    keys = [comp.key() for comp in compset.components]
    out: list[EquivalenceClass] = []
    for cl, _ in classes:
        everyone = [cl.representative] + cl.members
        everyone.sort(key=lambda c: _secondary_key(c, keys))
        cl.representative, cl.members = everyone[0], everyone[1:]
        out.append(cl)
    out.sort(key=lambda cl: cl.rss)
    return out


def _secondary_key(c: Candidate, comp_keys: list[tuple]) -> tuple:
    return (c.n_internal, 0 if c.sp else 1,
            str(c.network.serialize(comp_keys)))

File: AlgorithmLcr/Try2-Known component types, quantities and parameters/cppversion/src/selector.cpp
#include "selector.hpp"

#include <algorithm>
#include <cmath>
#include <map>

namespace ng {

std::vector<Candidate> evaluateCandidates(const std::vector<Network>& nets,
                                          const ComponentSet& compset,
                                          const std::vector<Complex>& s,
                                          const std::vector<Complex>& z,
                                          const std::vector<double>& w,
                                          int batchSize) {
    const int nObs = 2 * (int)z.size();
    const int p = compset.nParams();
    std::vector<Candidate> out;
    // group by structure, preserving first-appearance order (dict semantics)
    std::vector<std::pair<std::string, std::pair<Structure, std::vector<std::vector<int>>>>>
        byStructure;
    std::map<std::string, int> indexOf;
    for (const auto& net : nets) {
        std::string key = net.structure.key();
        auto it = indexOf.find(key);
        if (it == indexOf.end()) {
            indexOf[key] = (int)byStructure.size();
            byStructure.push_back({key, {net.structure, {net.assign}}});
        } else {
            byStructure[it->second].second.second.push_back(net.assign);
        }
    }
    for (auto& [key, entry] : byStructure) {
        const Structure& structure = entry.first;
        StructureStamps stamps = StructureStamps::build(structure, compset);
        bool spFlag = isSeriesParallel(structure.V, structure.mult);
        for (size_t i0 = 0; i0 < entry.second.size(); i0 += (size_t)batchSize) {
            size_t i1 = std::min(entry.second.size(), i0 + (size_t)batchSize);
            std::vector<std::vector<int>> chunk(entry.second.begin() + i0,
                                                entry.second.begin() + i1);
            std::vector<std::vector<Complex>> zModel = stamps.zFull(chunk, s);
            for (size_t row = 0; row < chunk.size(); ++row) {
                const std::vector<Complex>& zf = zModel[row];
                bool finite = true;
                for (const Complex& v : zf)
                    if (!std::isfinite(v.real()) || !std::isfinite(v.imag())) {
                        finite = false;
                        break;
                    }
                if (!finite) continue;
                std::vector<double> res = residualVector(z, zf, w);
                double rss = rssOf(res);
                auto [wrmse, mre] = fitMetrics(z, zf);
                Candidate c;
                c.network = Network{structure, chunk[row]};
                c.zFit = zf;
                c.rss = rss;
                c.aiccVal = aicc(rss, nObs, p);
                c.wrmse = wrmse;
                c.maxRelErr = mre;
                c.sp = spFlag;
                out.push_back(std::move(c));
            }
        }
    }
    std::stable_sort(out.begin(), out.end(),
                     [](const Candidate& a, const Candidate& b) { return a.rss < b.rss; });
    return out;
}

std::vector<double> makeValidationGrid(const std::vector<double>& f, int n,
                                       double expand) {
    double fmin = *std::min_element(f.begin(), f.end());
    double fmax = *std::max_element(f.begin(), f.end());
    // numpy.logspace over [log10(fmin/expand), log10(fmax*expand)]
    double la = std::log10(fmin / expand), lb = std::log10(fmax * expand);
    std::vector<double> out((size_t)n);
    for (int i = 0; i < n; ++i)
        out[i] = std::pow(10.0, la + (lb - la) * (double)i / (double)(n - 1));
    return out;
}

bool relDiffBelow(const std::vector<Complex>& za, const std::vector<Complex>& zb,
                  double tol) {
    double mx = -1.0;
    bool any = false;
    for (size_t k = 0; k < za.size(); ++k) {
        double denom = std::abs(za[k]);
        if (denom < 1e-300) denom = std::abs(zb[k]);
        double rel = std::abs(za[k] - zb[k]) / denom;
        if (!std::isfinite(rel)) continue;  // np.isfinite filter
        any = true;
        mx = std::max(mx, rel);
    }
    if (!any) return false;
    return mx < tol;
}

bool areEquivalent(const Network& a, const Network& b, const ComponentSet& compset,
                   const std::vector<double>& grid, double tol) {
    std::vector<Complex> za = networkZ(a, compset, grid);
    std::vector<Complex> zb = networkZ(b, compset, grid);
    return relDiffBelow(za, zb, tol);
}

namespace {

// str(network.serialize(comp_keys)) as the Python reference builds it --
// repr of the nested tuple (slot)(key)('C', value, dcr).  Reproduced here so
// the within-class representative choice matches the reference bit-for-bit
// (Python compares these strings, not the tuples).
std::string pySerialStr(const std::vector<std::vector<Component>>& serial) {
    auto keyStr = [](const Component& c) {
        return "('" + std::string(1, c.kind) + "', " + pyRepr(c.value) + ", " +
               pyRepr(c.dcr) + ")";
    };
    std::string out = "(";
    for (size_t k = 0; k < serial.size(); ++k) {
        if (k) out += ", ";
        out += "(";
        for (size_t q = 0; q < serial[k].size(); ++q) {
            if (q) out += ", ";
            out += keyStr(serial[k][q]);
        }
        if (serial[k].size() == 1) out += ",";
        out += ")";
    }
    if (serial.size() == 1) out += ",";
    out += ")";
    return out;
}

// (fewer internal junctions, series-parallel first, canonical wiring key)
struct SecondaryKey {
    int nInternal;
    int spPenalty;
    std::string serialStr;
    bool operator<(const SecondaryKey& o) const {
        if (nInternal != o.nInternal) return nInternal < o.nInternal;
        if (spPenalty != o.spPenalty) return spPenalty < o.spPenalty;
        return serialStr < o.serialStr;
    }
};

SecondaryKey secondaryKey(const Candidate& c, const std::vector<Component>& comps) {
    return SecondaryKey{c.nInternal(), c.sp ? 0 : 1, pySerialStr(c.network.serialize(comps))};
}

}  // namespace

std::vector<EquivalenceClass> rankAndCluster(std::vector<Candidate> candidates,
                                             const ComponentSet& compset,
                                             const std::vector<double>& f,
                                             int clusterTop, double equivTol) {
    if (candidates.empty()) return {};
    double sigmaHat = candidates[0].wrmse;  // candidates are RSS-sorted; py takes
    for (const auto& c : candidates)        // min(wrmse) -- same set, same min
        sigmaHat = std::min(sigmaHat, c.wrmse);
    double tol = std::max(equivTol, 3.0 * sigmaHat);
    std::vector<double> grid = makeValidationGrid(f);
    size_t topN = std::min((size_t)clusterTop, candidates.size());
    std::vector<std::vector<Complex>> zGrids(topN);
    for (size_t i = 0; i < topN; ++i)
        zGrids[i] = networkZ(candidates[i].network, compset, grid);

    std::vector<EquivalenceClass> classes;
    std::vector<std::vector<Complex>> classZ;
    for (size_t i = 0; i < topN; ++i) {
        bool matched = false;
        for (size_t ci = 0; ci < classes.size(); ++ci) {
            if (relDiffBelow(classZ[ci], zGrids[i], tol)) {
                classes[ci].members.push_back(std::move(candidates[i]));
                matched = true;
                break;
            }
        }
        if (!matched) {
            classes.push_back(EquivalenceClass{std::move(candidates[i]), {}});
            classZ.push_back(zGrids[i]);
        }
    }
    const auto& comps = compset.components();
    for (auto& cl : classes) {
        std::vector<Candidate> everyone;
        everyone.push_back(std::move(cl.representative));
        for (auto& m : cl.members) everyone.push_back(std::move(m));
        std::stable_sort(everyone.begin(), everyone.end(),
                         [&](const Candidate& a, const Candidate& b) {
                             return secondaryKey(a, comps) < secondaryKey(b, comps);
                         });
        cl.representative = std::move(everyone[0]);
        cl.members.clear();
        for (size_t i = 1; i < everyone.size(); ++i)
            cl.members.push_back(std::move(everyone[i]));
    }
    std::stable_sort(classes.begin(), classes.end(),
                     [](const EquivalenceClass& a, const EquivalenceClass& b) {
                         return a.rss() < b.rss();
                     });
    return classes;
}

}  // namespace ng
